From edc3d5c5316bc8078dc1ba09237f920ce411bef1 Mon Sep 17 00:00:00 2001
From: Jiri Pirko <jpirko@redhat.com>
Date: Tue, 6 May 2014 09:15:32 -0400
Subject: [net] ip_tunnel: Set network header properly for IP_ECN_decapsulate()

Message-id: <1399367732-13743-1-git-send-email-jpirko@redhat.com>
Patchwork-id: 79404
O-Subject: [patch rhel7 net] ip_tunnel: Set network header properly for IP_ECN_decapsulate()
Bugzilla: 1094419
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Veaceslav Falico <vfalico@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

BZ1094419
https://brewweb.devel.redhat.com/taskinfo?taskID=7419892

upstream (net) commit e96f2e7c430014eff52c93cabef1ad4f42ed0db1
Author: Ying Cai <ycai@google.com>
Date:   Sun May 4 15:20:04 2014 -0700

    ip_tunnel: Set network header properly for IP_ECN_decapsulate()

    In ip_tunnel_rcv(), set skb->network_header to inner IP header
    before IP_ECN_decapsulate().

    Without the fix, IP_ECN_decapsulate() takes outer IP header as
    inner IP header, possibly causing error messages or packet drops.

    Note that this skb_reset_network_header() call was in this spot when
    the original feature for checking consistency of ECN bits through
    tunnels was added in eccc1bb8d4b4 ("tunnel: drop packet if ECN present
    with not-ECT"). It was only removed from this spot in 3d7b46cd20e3
    ("ip_tunnel: push generic protocol handling to ip_tunnel module.").

Signed-off-by: Jiri Pirko <jpirko@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/ip_tunnel.c b/net/ipv4/ip_tunnel.c
index 7116f96..94a70be 100644
--- a/net/ipv4/ip_tunnel.c
+++ b/net/ipv4/ip_tunnel.c
@@ -436,6 +436,8 @@ int ip_tunnel_rcv(struct ip_tunnel *tunnel, struct sk_buff *skb,
   tunnel->i_seqno = ntohl(tpi->seq) + 1;
  }
 
+ skb_reset_network_header(skb);
+
  err = IP_ECN_decapsulate(iph, skb);
  if (unlikely(err)) {
   if (log_ecn_error)
-- 
1.7.1